### ISSUE TITLE
Stop pinning everything in CI

### ISSUE THREAD

**Issue by LLFourn** [bitcoindevkit/bdk#1398]
Why do we need all these pins?

We don't have to test on MSRV we just have to check that it builds on MSRV. dev-dependencies shouldn't have to be pinned. Or at least that's the policy I'd like to adopt.

See sophisticated technology on how to do this: https://github.com/bitcoindevkit/coin-select/blob/990226a982bd6036bed5a7674fe57d6f9d48e49e/.github/workflows/test.yml#L40

cc @notmandatory

_Originally posted by @LLFourn in https://github.com/bitcoindevkit/bdk/pull/1397#discussion_r1553437195_

**Comment by LLFourn** [comment]
> > We don't have to test on MSRV we just have to check that it builds on MSRV. dev-dependencies shouldn't have to be pinned.
> 
> I disagree: MSRV means "This project supports/works with this rust version", and, everything untested may be considered broken. I regularly run into MSRV breakages for depedencies that 'check' but don't test their MSRV on all supported platforms.

This "regularly" is quite concerning tbh :sweat_smile:. It's a good point that if our tests don't have the same MSRV as build then there's no way to verify that the tests actually pass on the MSRV. I want to add that this situation would imply a bug in the rust compiler, either at the version in CI or at the MSRV. Either that or some weird rust version based conditional compilation in a dependency.

For me the monotonically growing list of dependency pins and using old versions of dev dependencies is a worse problem than this. I definitely don't take the attitude that "everything untested may be considered broken" when developing in rust. Logic that is not tested can be considered broken but not *everything* :)

Not a hill I'll die on though and if the maintainers who are really affected by this want to take the stricter interpretation of MSRV that's fine with me!

**Comment by notmandatory** [comment]
Since updating MSRV to 1.63 we were able to remove many of the originally pinned versions in CI. As 1.63 gets older this list will likely grow again and we'll need to re-verify they're still needed but otherwise I don't see any other way to reduce the pins except bumping our MSRV at least when ever a new Debian stable release comes out with an updated Rust version.

**Comment by evanlinjin** [comment]
I don't think we need to make sure tests run under MSRV, it just makes testing unnecessarily difficult. We can't use the latest tooling. There is no workaround to this because dev-dependencies cannot be optional.

**Comment by notmandatory** [comment]
I'm open to taking this approach but would like to know if there are any other big projects who handle CI MSRV maintenance this way? If we update our CI to not use the automatic binary download feature of `bitcoind` and `electrsd` (install locally in a custom docker image instead) I think we can eliminate a bunch of the troublesome dev dependencies.

**Comment by notmandatory** [comment]
Can we move this out of the 1.0.0 milestone since it will require some discussion and only affects CI?

**Comment by notmandatory** [comment]
I think all we need to do to fix this is exclude `bdk_testenv` when doing the MSRV build and don't try to run the tests when using the MSRV. I made these changes in #1776 and only 6 non-test related dependencies needed to be pinned.

**Comment by notmandatory** [comment]
> I think all we need to do to fix this is exclude `bdk_testenv` when doing the MSRV build and don't try to run the tests when using the MSRV. I made these changes in #1776 and only 6 non-test related dependencies needed to be pinned.

I removed this change from 1776 since maybe it's not a good idea to not run tests with MSRV also. See also: https://github.com/bitcoindevkit/bdk/pull/1776#pullrequestreview-2507362798
